### PR TITLE
[GraphBolt] replace item<> with raw pointer

### PR DESCRIPTION
**Pull request by Rhett-Ying**
## Description

wrapped with extra `AT_DISPATCH_INTEGRAL_TYPES`.
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Review by mfbalin — APPROVED** [review]
LGTM.

File: graphbolt/src/fused_csc_sampling_graph.cc [hunk 1]
```diff
@@ -330,96 +330,112 @@ FusedCSCSamplingGraph::SampleNeighborsImpl(
   torch::optional<torch::Tensor> subgraph_type_per_edge = torch::nullopt;
 
   AT_DISPATCH_INTEGRAL_TYPES(
-      indptr_.scalar_type(), "SampleNeighborsImpl", ([&] {
-        const scalar_t* indptr_data = indptr_.data_ptr<scalar_t>();
-        auto num_picked_neighbors_data_ptr =
-            num_picked_neighbors_per_node.data_ptr<scalar_t>();
-        num_picked_neighbors_data_ptr[0] = 0;
+      indptr_.scalar_type(), "SampleNeighborsImplWrappedWithIndptr", ([&] {
```
Comment (peizhou001) [inline graphbolt/src/fused_csc_sampling_graph.cc:333]: Don't change name?

File: graphbolt/src/fused_csc_sampling_graph.cc [hunk 2]
```diff
@@ -330,96 +330,112 @@ FusedCSCSamplingGraph::SampleNeighborsImpl(
...
+        using indptr_t = scalar_t;
```
Comment (peizhou001) [inline graphbolt/src/fused_csc_sampling_graph.cc:334]: why need this

File: graphbolt/src/fused_csc_sampling_graph.cc (hunk 1 above)
Comment (Rhett-Ying) [reply graphbolt/src/fused_csc_sampling_graph.cc:333]: just to be consistent with following wrapper with `nodes`. Any concern about this name? any other dependency on this name beside its uniqueness requirement?

**Comment by peizhou001** [comment]
The change makes sense, but one thing needs discussion is do we need to carefully deal with the type nodes? or just do a explicit conversion in python side at tensor level, or just do implicit conversion in C side at variable level?

**Comment by Rhett-Ying** [comment]
> The change makes sense, but one thing needs discussion is do we need to carefully deal with the type nodes? or just do a explicit conversion in python side at tensor level, or implicit conversion in C side at variable level?

The main purpose I added an extra dispatch macro is to avoid conversion. If conversion is applied always that makes sure `nodes.dtype` equals `indices.dtype`, that is the safest. Or user is responsible for indexing on graph with `nodes` IDs is dtype-safe.

**Review by frozenbugs — APPROVED** [review]

**Pull request merged**